### ISSUE TITLE
ImGui::IsItemHovered() and mouse clicks

### ISSUE THREAD

**Issue by Flix01** [ocornut/imgui#1013]
In one of my projects I have a very strange behaviour: when I keep the left or the middle mouse button pressed, ImGui::IsItemHovered() becomes false (even if I'm on the item). Right clicks don't seem to be affected by this issue, and it always happens regardless the ImGui back-end (glfw3 or SDL2).

Basically I want to process LMB and MMB clicks when the user is on an item, and I can't, because when the mouse is down,  ImGui::IsItemHovered() becomes false.

I know, it's probably just something wrong in my code, and to have better answers I should attach some code to replicate the issue, but can anybody guess what's happening ?

**P.S.** I know that I could use _ImGui::IsMouseHoveredRect()_ to make it work, and it works this way, but then _I'm afraid it will trigger even on windows that are covered by the active one_...

**Comment by Flix01** [comment]
Closing here... I'm on my way to solve it.

**Comment by Flix01** [comment]
Basically I've spot in my code something like:
```c++
if (!ImGui::ItemAdd(bb, NULL)) return rv;
```
And I'm not sure I can leave the ID to NULL.

**Comment by Flix01** [comment]
Nope, it was not ImGui::ItemAdd(...). :disappointed:

**Comment by Flix01** [comment]
For now I'm using:
```c++
const bool itemHovered = ImGui::IsItemHoveredRect() && ImGui::IsWindowHovered();
 ```
Hope it works pretty much the same way as:
```c++
const bool itemHovered = ImGui::IsItemHovered();
```

**Comment by Flix01** [comment]
This is the code I've made to reproduce the issue:
```c++
#include <imgui.h>	
#include <imgui_internal.h>

void DrawGL()	
{
//	[...]
        static bool open = true;
        if (ImGui::Begin("Debug Window", &open, ImVec2(600,600))) {

        if (ImGui::BeginChild("Test IsItemHovered()")) {
            const ImGuiIO& io = ImGui::GetIO();

            ImGuiWindow* window = ImGui::GetCurrentWindow();
            if (window && !window->SkipItems) {
                const ImVec2 wndSz = ImGui::GetWindowSize();
                ImRect bb(window->DC.CursorPos, ImVec2(window->DC.CursorPos.x + wndSz.x,window->DC.CursorPos.y + wndSz.y));
                ImGui::ItemSize(bb);
                if (ImGui::ItemAdd(bb, NULL)) {
                    ImGui::SetTooltip("IsItemHovered() = %s\nLMB down = %s\nMMB down = %s\nRMB down = %s",
                                      ImGui::IsItemHovered()?"true":"false",
                                      io.MouseDown[0]?"true":"false",
                                      io.MouseDown[2]?"true":"false",
                                      io.MouseDown[1]?"true":"false"
                                      );
                }
            }
        }
        ImGui::EndChild();
        }
        ImGui::End();
}
```

When I run this code and I click the _Left Mouse Button_, _IsItemHovered()_ becomes false.

**Comment by colesnicov** [comment]
Hi Flix01.
Yes, i have same problem here: https://github.com/ocornut/imgui/issues/1014 when i clicked on tracker the component lost hovered state...

To you problem. I'm not sure if is resolve your problem but if you know about function `ImGui::ButtonBehavior`

**Comment by Flix01** [comment]
@colesnicov 
Yes, getting itemHovered from ImGui::ButtonBehavior(...) seems to fix things. 
Thanks for your suggestion! :smiley: 

However if I use it, I got another issue that affects only my glfw3 back-end (the SDL2 one seems to work fine as far as I can see): the middle mouse button clicks are not detected correctly (although left clicks now work).

I'm not sure why this happen. However now I got this 2 alternatives:
```c++
const bool itemHovered = ImGui::IsItemHoveredRect() && ImGui::IsWindowHovered();
```
and
```c++
bool itemHovered = false;
// Needs the ImRect (bb) and the ID (id):
ImGui::ButtonBehavior(bb,id,&itemHovered,NULL);
```
I think I can be happy now.
